### ISSUE TITLE
Resources without Resx

### ISSUE THREAD

**Issue by bchavez** [nikhilk/scriptsharp#115]
Hi,

I was hoping to generate resource types without having to create a Resx file.

I'm trying to achieve the following output from ScriptSharp (a global variable for a utility class):

``` javascript
PhotoSize = { 
    Crop: 'c',
    Small: 's',
    Normal: 'n',
    Square: 'q'
};
```

A `new Dictionary("key","value",...)` pattern would work, but there's no way to define a "global variable" at the assembly level in C#.

So, I found that if I do something similar to:

```
[Resources]
public static class PhotoSize
{
    [PreserveCase]
    public static readonly string Crop = "c";
    [PreserveCase]
    public static readonly string Small = "s";
    [PreserveCase]
    public static readonly string Normal = "n";
    [PreserveCase]
    public static readonly string Square = "q";
}
```

I could get close... however, the output was:

``` javascript
PhotoSize = { 
    Crop: '',
    Small: '',
    Normal: '',
    Square: ''
};
```

without any initialization. :(

Resx resource strings work exactly as intended; however, I'd like to explicitly define them without having to create a Resx file.

Your thoughts?

I'm hoping we could take it one step further to allow explicit variable initialization.

Thanks,
Brian

**Comment by Griffitsj** [comment]
If you make the fields const instead of static readonly it produces the output you desire.

<pre><code>[Resources]
public class PhotoSize
{
    [PreserveCase]
    public const string Crop = "c";
    [PreserveCase]
    public const string Small = "s";
    [PreserveCase]
    public const string Normal = "n";
    [PreserveCase]
    public const string Square = "q";
}
</code></pre>


You will also want to add the [IgnoreNamespace] attribute to the class.

**Comment by bchavez** [comment]
You are a ScriptSharp genius... thank you Griffitsj.

Should I keep this issue open, even though `readonly` is a valid construct too?

**Comment by Griffitsj** [comment]
I think that the issue should be kept open because const can not be used for all types.  DateTime is an example.

**Comment by nikhilk** [comment]
You can also used an enum to define a set of named values.

```
[NamedValues]
public enum PhotoSize {
    [ScriptName("c")] Crop,
    [ScriptName("s")] Small,
    [ScriptName("n")] Normal,
    [ScriptName("q")] Square
}
```

This will actually allow you to refer to them in code as PhotoSize.Crop etc. and allow you to strongly type variables, rather than typing them as string. The generated code will contain literal strings.

If you make the enum public, an actual object with these values will also be defined.

**Comment by bchavez** [comment]
Hi Nikhil,

I tried your suggestion, but didn't get the desired output, here's what I got:

```
// C#
[NamedValues]
public enum PhotoSize
{
    [ScriptName( "c" )]
    Crop = 1,
    [ScriptName( "q" )]
    Square = 2,
    [ScriptName( "s" )]
    Small = 3,
    [ScriptName( "n" )]
    Normal = 4,
    [ScriptName( "l" )]
    Large = 5
}
// Note: Got an error "Enumeration fields must have an explicit constant value specified"
//       if explicit enum values were not defined. 

//JavaScript
PhotoSize2.prototype = {
    c: 'c', 
    q: 'q', 
    s: 's', 
    n: 'n', 
    l: 'l'
}
PhotoSize2.registerEnum('PhotoSize2', false);
```

From a purely Script# perspective, true, I'll be able to PhotoSize.Crop from my ScriptSharp C# code. However, if I'm accessing this enum from "outside" of Script#, say, in a `<script>` tag in a webpage, I would have to pass `PhotoSize.c` to all of my utility methods that require a PhotoSize:

```
PhotoUtil.GetPhoto("... url ...", PhotoSize.c);
// V.S.
PhotoUtil.GetPhoto("... url ...", PhotoSize.Crop);
```

I'd prefer to use `PhotoSize.Crop` instead of `PhotoSize.c` for readability & usability of my API.

My first natural inclination was first to try the following:

```
[CharValues]
public enum PhotoSize
{
    Crop = 'c',
    Square = 'q',
    Small = 's',
    Normal = 'n',
    Large = 'l'   
}
```

I was digging around Script# for something like `CharValues` that could compliment `NamedValues`; but didn't find anything. I need to have "c", "q", "s", "n", and "l" characters preserved in the enum because these values are used in the calculation a photo URL.

I think including something like `[CharValues]` might be a good approach. No?

(similar to NHibernate's handling of enums with `EnumCharType` and `EnumStringType`).

Thanks for everyone's help.

**Comment by nikhilk** [comment]
Interesting point - I agree, for the public enum, you want the value to be 'c' and the name of the field to still be "Crop". I think [NamedValues] should just do it, rather than invent another attribute.

That said, usually native script APIs usually document sending in literal values, if you look at general patterns and precedents... so the caller would be passing in 'c' and not PhotoSize.c (or PhotoSize.Crop). However I still agree, if you want to create constants as part of your API surface, their naming should reflect actual intent.

**Comment by nikhilk** [comment]
Actually in looking further at this, I found it hard to decouple the name of a named value enum to the value itself. However I noticed there shouldn't be a real problem here... I was able to generate the same end-result...

Here is the c# source:

```
public static class PhotoSize
{
    [PreserveCase]
    public static readonly string Crop = "c";
    [PreserveCase]
    public static readonly string Small = "s";
    [PreserveCase]
    public static readonly string Normal = "n";
    [PreserveCase]
    public static readonly string Square = "q";
}
```

And here is the generated script:

```
App.PhotoSize = function() {
}
App.PhotoSize.registerClass('App.PhotoSize');
App.PhotoSize.Crop = 'c';
App.PhotoSize.Small = 's';
App.PhotoSize.Normal = 'n';
App.PhotoSize.Square = 'q';
```

If you don't want this to be namespace'd ... i.e. you really want to create a global, you could add [IgnoreNamespace] or potentially even [ScriptNamespace("window")] if you want to export it out of the scope representing your script.

Basically, I don't think you'd need to resort to [Resources] and trying to bend the infrastructure for resx to this scenario.

Given this, I'll close out the bug at this point.

Yes, there is potentially an optimization to be made around declaring static members when there are multiple members to be defined, so you get the exact syntax that you first listed as well.
